### PR TITLE
CORTX-31409 Jenkins job for whitesource container scan

### PR DESCRIPTION
**Pull request by Abhirup-A-G**
Signed-off-by: Abhirup Ghosh <abhirup.ghosh@seagate.com>

# Problem Statement
- CORTX-31409 Jenkins job for whitesource container scan

# Design
-  Script Validation.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM
- [ ] Jenkins pipelines used for testing

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: jenkins/security/whitesource_container_scan.groovy [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'CONTROL_EXTERNAL_NODEPORT', defaultValue: '31169', description: 'Port to be used for control service.', trim: true)
+        string(name: 'S3_EXTERNAL_HTTP_NODEPORT', defaultValue: '30080', description: 'HTTP Port to be used for IO service.', trim: true)
+        string(name: 'S3_EXTERNAL_HTTPS_NODEPORT', defaultValue: '30443', description: 'HTTPS to be used for IO service.', trim: true)
+        text(defaultValue: '''hostname=<hostname>,user=<user>,pass=<password>''', description: 'VM details to be used. First node will be used as Primary node', name: 'hosts')
+        choice(
+            name: 'EXTERNAL_EXPOSURE_SERVICE',
+            choices: ['LoadBalancer', 'NodePort'],
+            description: 'K8s Service to be used to expose RGW Service to outside cluster.'
+        )
+        string(name: 'WHITESOURCE_SERVER_URL', defaultValue: ' ', description: 'WhiteSource Server URL', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:42]: Please add a default value

File: jenkins/security/whitesource_container_scan.groovy [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'API_KEY', defaultValue: ' ', description: 'apiKey', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:43]: Please add a default value

File: jenkins/security/whitesource_container_scan.groovy [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'USER_KEY', defaultValue: ' ', description: 'userKey', trim: true)
+        string(name: 'PRODUCT_NAME', defaultValue: ' ', description: 'productName', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:45]: Please add a default value

File: jenkins/security/whitesource_container_scan.groovy [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'DOCKER_REGISTRY', defaultValue: ' ', description: 'docker registry', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:46]: Please add a default value

File: jenkins/security/whitesource_container_scan.groovy [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'MAIN_POD', defaultValue: ' ', description: 'mainPod details', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:47]: Please add a default value. Do we need this value as input? Maybe we can fetch these details from the primary node.

File: jenkins/security/whitesource_container_scan.groovy [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'WORKER_POD', defaultValue: ' ', description: 'workerPod details', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:48]: Please add a default value. Do we need this value as input? Maybe we can fetch these details from the primary node.

File: jenkins/security/whitesource_container_scan.groovy [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        string(name: 'PULL_SECRET', defaultValue: ' ', description: 'image pullSecretL', trim: true)
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:49]: We can is this from Jenkins credentials. Refer - https://github.com/Seagate/cortx-re/blob/main/jenkins/internal-ci/generic/branch/cortx-images.groovy#L17

File: jenkins/security/whitesource_container_scan.groovy [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+                        export SOLUTION_CONFIG_TYPE=automated
+                        export WHITESOURCE_SERVER_URL=${WHITESOURCE_SERVER_URL}
+                        export API_KEY=${API_KEY}
+                        export USER_KEY=${USER_KEY}
+                        export PRODUCT_NAME=${PRODUCT_NAME}
+                        export DOCKER_REGISTRY=${DOCKER_REGISTRY}
+                        export MAIN_POD=${MAIN_POD}
+                        export WORKER_POD=${WORKER_POD}
+                        export PULL_SECRET=${PULL_SECRET}
+                        ./whitesource_container_scan.sh
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:120]: Do we need any change in ./whitesource_container_scan.sh. If so please add to this PR only.

File: jenkins/security/whitesource_container_scan.groovy [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    post {
+        always {
+
+            script {
+
+                // Jenkins Summary
+                clusterStatus = ""
+                if ( fileExists('/var/tmp/cortx-cluster-status.txt') && currentBuild.currentResult == "SUCCESS" ) {
+                    clusterStatus = readFile(file: '/var/tmp/cortx-cluster-status.txt')
+                    MESSAGE = "CORTX Cluster Setup Success for the build ${build_id}"
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:136]: Please update the email notification message for the WhiteSource scan

File: jenkins/security/whitesource_container_scan.groovy [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+                    ICON = "accept.gif"
+                    STATUS = "SUCCESS"
+                } else if ( currentBuild.currentResult == "FAILURE" ) {
+                    manager.buildFailure()
+                    MESSAGE = "CORTX Cluster Setup Failed for the build ${build_id}"
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:141]: Please update the email notification message for the WhiteSource scan

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

File: jenkins/security/whitesource_container_scan.groovy (hunk 1 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:42]: Done

File: jenkins/security/whitesource_container_scan.groovy (hunk 2 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:43]: Done

File: jenkins/security/whitesource_container_scan.groovy (hunk 3 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:45]: Done

File: jenkins/security/whitesource_container_scan.groovy (hunk 4 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:46]: Done

File: jenkins/security/whitesource_container_scan.groovy (hunk 6 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:48]: As Discussed Swanand want the mainPod & workerpod to be parameterized. As only the vision will be a variable hence removing mainPod and worker pod variable and keeping only one as whitesource_version

File: jenkins/security/whitesource_container_scan.groovy (hunk 5 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:47]: As Discussed Swanand want the mainPod & workerpod to be parameterized. As only the vision will be a variable hence removing mainPod and worker pod variable and keeping only one as whitesource_version

File: jenkins/security/whitesource_container_scan.groovy (hunk 8 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:120]: Ok

File: jenkins/security/whitesource_container_scan.groovy (hunk 1 above)
Comment (shailesh-vaidya) [reply jenkins/security/whitesource_container_scan.groovy:42]: I still see `default value` as blank 
`string(name: 'WHITESOURCE_SERVER_URL', defaultValue: ' ', description: 'WhiteSource Server URL', trim: true)`
Could you please check

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

File: jenkins/security/whitesource_container_scan.groovy [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    
+    options {
+        timeout(time: 240, unit: 'MINUTES')
+        timestamps()
+        buildDiscarder(logRotator(daysToKeepStr: '30', numToKeepStr: '30'))
+        ansiColor('xterm')
+    }
+
+    environment {
+        PODS_ON_PRIMARY = "true"
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:16]: Do we need to allow POD creation on the Primary node? I guess we are planning to create a 1+3 setup. In that case this should be `false` 
@swanand-gadre - Could you please confirm.

File: jenkins/security/whitesource_container_scan.groovy (hunk 7 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:49]: Implemented it

File: jenkins/security/whitesource_container_scan.groovy (hunk 9 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy]: Removed this section, as its not required

File: jenkins/security/whitesource_container_scan.groovy (hunk 10 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy]: Removed this section, as its not required

File: jenkins/security/whitesource_container_scan.groovy (hunk 7 above)
Comment (shailesh-vaidya) [reply jenkins/security/whitesource_container_scan.groovy:49]: I still see `PULL_SECRET` as the input variable. This need to be removed and fetched from Jenkins credentials.

File: jenkins/security/whitesource_container_scan.groovy [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                    pushd solutions/kubernetes/
+                        echo $hosts | tr ' ' '\n' > hosts
+                        cat hosts
+                        if [ "$(cat hosts | wc -l)" -eq 2 ]
+                        then
+                           echo "Current configuration does not support 2 node CORTX cluster deployment. Please try with 1 or more than two nodes."
+                           echo "Exiting Jenkins job."
+                           exit 1
+                        fi
+                        ./cluster-setup.sh ${PODS_ON_PRIMARY}
```
Comment (shailesh-vaidya) [inline jenkins/security/whitesource_container_scan.groovy:77]: Do we need `${PODS_ON_PRIMARY}` if we are always planing to use 3+1 setup

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

File: jenkins/security/whitesource_container_scan.groovy (hunk 12 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:77]: I have seen your previous comment, as we are yet to receive the update from swanand so i didn't took any action on it. if PODS_ON_PRIMARY = "false" do we still need to declare it explicitly? @shailesh-vaidya

Comment (shailesh-vaidya) [reply jenkins/security/whitesource_container_scan.groovy:77]: We do not need to declare it if it's false. Please follow up with @swanand-gadre and get it confirmed.

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

File: jenkins/security/whitesource_container_scan.groovy (hunk 12 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:77]: removed ${PODS_ON_PRIMARY}

File: jenkins/security/whitesource_container_scan.groovy (hunk 11 above)
Comment (Abhirup-A-G) [reply jenkins/security/whitesource_container_scan.groovy:16]: removed ${PODS_ON_PRIMARY}

**Comment by shailesh-vaidya** [comment]
> Thanks for your contribution! The CLA bot has flagged your contribution as not having a Contributor License Agreement in place. Note that this is not needed in the overwhelming majority of instances and this warning will usually be ignored. The code reviewers will make a determination and may ask you to sign a CLA or may choose to ignore this warning. More information about this can be found [here](https://github.com/Seagate/cortx/blob/master/doc/dco_cla.md).

@Abhirup-A-G  Seems cla signing is failing for this PR. Please check with @mukul-seagate11

File: jenkins/security/whitesource_container_scan.groovy (hunk 11 above)
Comment (shailesh-vaidya) [reply jenkins/security/whitesource_container_scan.groovy:16]: it's still there https://github.com/Abhirup-A-G/cortx-re/blob/CORTX-31409/jenkins/security/whitesource_container_scan.groovy#L16

File: scripts/security/whitesource_container_scan.sh [hunk 13]
```diff
@@ -33,14 +34,16 @@ echo -e "\n\n###################################################################
...
 }
 default_parameter
+
+pushd ws-k8s-agent/helm-chart/
 #Updating the configuration file
-sed -Ei 's,(url: ).*,\1"'"$WHITESOURCE_SERVER_URL"'",g' $AGENT_TAR_LOCATION; sed -Ei 's,(apiKey: ).*,\1"'"$API_KEY"'",g' $AGENT_TAR_LOCATION
-sed -Ei 's,(userKey: ).*,\1"'"$USER_KEY"'",g' $AGENT_TAR_LOCATION; sed -Ei "s,(productName: ).*,\1$PRODUCT_NAME,g" $AGENT_TAR_LOCATION
-sed -Ei "s,(registry: ).*,\1$DOCKER_REGISTRY,g" $AGENT_TAR_LOCATION; sed -Ei "s,(mainPod: ).*,\1$MAIN_POD,g" $AGENT_TAR_LOCATION
-sed -Ei "s,(workerPod: ).*,\1$WORKER_POD,g" $AGENT_TAR_LOCATION; sed -Ei 's,(pullSecret: ).*,\1"'"$PULL_SECRET"'",g' $AGENT_TAR_LOCATION
+sed -Ei 's,(url: ).*,\1"'"$WHITESOURCE_SERVER_URL"'",g' values.yaml; sed -Ei 's,(apiKey: ).*,\1"'"$API_KEY"'",g' values.yaml
```
Comment (shailesh-vaidya) [inline scripts/security/whitesource_container_scan.sh:40]: Please check if we can optimize this logic using loop

**Review by shailesh-vaidya — CHANGES_REQUESTED** [review]

**Pull request closed without merging**